CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc

```diff
@@ -567,35 +569,65 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
 		// Don't allow non hero units to purchase abilities with AP without a training center
 		// The branch is checked here so that "No Training Center" takes priority over "No
 		// class perk picked" as a reason.
-		if((UnitState.HasPurchasedPerkAtRank(Rank) || Branch >= AbilityRanks) && !UnitState.IsResistanceHero() && !CanSpendAP())
+		if ((UnitState.HasPurchasedPerkAtRank(Rank) || Branch >= AbilityRanks) && !UnitState.IsResistanceHero() && !CanSpendAP())
 		{
-			ReasonLockedText = ReasonNoTrainingCenter;
-			return false;
-		}
-
-		// Don't allow non hero units to purchase abilities on the xcom perk row before getting a rankup perk
-		if(!UnitState.HasPurchasedPerkAtRank(Rank) && !UnitState.IsResistanceHero() && Branch >= AbilityRanks )
+			strLocReasonLocked = ReasonNoTrainingCenter;
+			nReasonLocked = 'NoTrainingCenter';
+		} // Don't allow non hero units to purchase abilities on the xcom perk row before getting a rankup perk
+		else if (!UnitState.HasPurchasedPerkAtRank(Rank) && !UnitState.IsResistanceHero() && Branch >= AbilityRanks )
 		{
-			ReasonLockedText = ReasonNoClassPerkPurchased;
-			return false;
+			strLocReasonLocked = ReasonNoClassPerkPurchased;
+			nReasonLocked = 'NoClassPerkPurchased';
 		}
-	}
-
-	// Normal behaviour and hero promotion emulation (ability ranks == 0)
-	if (Rank >= UnitState.GetRank())
+	} // Normal behaviour and hero promotion emulation (ability ranks == 0)
+	else if (Rank >= UnitState.GetRank())
 	{
-		ReasonLockedText = ReasonNotHighEnoughRank;
+		strLocReasonLocked = ReasonNotHighEnoughRank;
+		nReasonLocked = 'NotHighEnoughRank';
 	}
 	else if (!CanAffordAbility(Rank, Branch))
 	{
-		ReasonLockedText = ReasonNotEnoughAP;
+		strLocReasonLocked = ReasonNotEnoughAP;
+		nReasonLocked = 'NotEnoughAP';
 	}
 	else if (!UnitState.MeetsAbilityPrerequisites(AbilityName))
 	{
-		ReasonLockedText = ReasonLacksPrerequisites;
+		strLocReasonLocked = ReasonLacksPrerequisites;
+		nReasonLocked = 'LacksPrerequisites';
```

Comment:
Perhaps these should be constants so that mod authors won't be tripped up by mistyping one of these names.

Reply:
Mods wouldn't be able to access those constants without building against CPS, which I don't think is warranted for something like this. Otherwise I would have made it a `enum`. We can still keep them as constants locally, if you think it's warranted, but they would not be used anywhere outside this one function anyway, so I don't think there's point in doing that.

Reply:
Good point. I didn't think about that. Yeah, we can't make it a requirement to build against the promotion screen.